lineoperations/src/linefunctions.c

```diff
@@ -0,0 +1,401 @@
+/*
+ *      linefunctions.c - Line operations, remove duplicate lines, empty lines, lines with only whitespace, sort lines.
+ *
+ *      Copyright 2015 Sylvan Mostert <smostert.dev@gmail.com>
+ *
+ *      This program is free software; you can redistribute it and/or modify
+ *      it under the terms of the GNU General Public License as published by
+ *      the Free Software Foundation; either version 2 of the License, or
+ *      (at your option) any later version.
+ *
+ *      This program is distributed in the hope that it will be useful,
+ *      but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *      GNU General Public License for more details.
+ *
+ *      You should have received a copy of the GNU General Public License along
+ *      with this program; if not, write to the Free Software Foundation, Inc.,
+ *      51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+
+#include "linefunctions.h"
+
+
+// isspace()
+gboolean issp(gchar c)
+{
+	return (c == ' ' || c == '\t' || c == '\f' ||
+			c == '\v' || c == '\r' || c == '\n');
+}
+
+
+// iswhitespaceline(gchar* line) checks if line is a whitespace line
+gboolean iswhitespaceline(gchar* line)
+{
+	gint i;
+
+	for(i = 0; line[i] != '\0'; i++)
+		if(!issp(line[i]))
+			return FALSE;
+	return TRUE;
+}
+
+
+// comparison function to be used in qsort
+static gint compareasc(const void * a, const void * b)
+{
+	return strcmp(*(const gchar **) a, *(const gchar **) b);
+}
+
+
+// comparison function to be used in qsort
+static gint comparedesc(const void * a, const void * b)
+{
+	return strcmp(*(const gchar **) b, *(const gchar **) a);
+}
+
+
+// Remove Duplicate Lines, sorted
+void rmdupst(GeanyDocument *doc) {
+	gint  totalnumchars;    // number of characters in the document
+	gint  totalnumlines;    // number of lines in the document
+	gchar **lines;          // array to hold all lines in the document
+	gint  numlines;         // number of lines in **lines array
+	gchar *newfile;         // *final* string to replace current document
+	gint  nfposn;           // keeps track of the position of newfile string
+	gint  i;                // iterator
+	gint  j;                // iterator
+	gint  k;                // iterator
+
+	totalnumchars = sci_get_length(doc->editor->sci);
+	totalnumlines = sci_get_line_count(doc->editor->sci);
+	lines         = g_malloc(sizeof(gchar *) * totalnumlines);
+	newfile       = g_malloc(sizeof(gchar) * (totalnumchars+1));
+	numlines      = 0;
+	nfposn        = 0;
+	k             = 0;
+
+	if(newfile && lines)    // verify memory allocation
```

Comment:
Thanks, this is now updated. I decided to keep g_malloc and remove the null check since it is unnecessary  as "GLib/GTK/Geany will most likely abort anyways.